platform/iphone/gl_view.mm

```diff
@@ -379,369 +222,106 @@ - (void)destroyFramebuffer {
 }
 
 - (void)startAnimation {
-	if (active)
-		return;
-	active = TRUE;
-	printf("start animation!\n");
-	if (useCADisplayLink) {
+	// Already active?
+	if (self.isActive) return;
+
+	self.active = YES;
 
-		// Approximate frame rate
-		// assumes device refreshes at 60 fps
-		int frameInterval = (int)floor(animationInterval * 60.0f);
+	printf("start animation!\n");
 
-		displayLink = [CADisplayLink displayLinkWithTarget:self selector:@selector(drawView)];
-		[displayLink setFrameInterval:frameInterval];
+	if (self.useCADisplayLink) {
+
+		// Create a display link and try and grab the fatest refresh rate
+		_displayLink = [CADisplayLink displayLinkWithTarget:self
+												   selector:@selector(drawView)];
+		if ([_displayLink respondsToSelector:@selector(preferredFramesPerSecond)]) {
+			_displayLink.preferredFramesPerSecond = 0;
+			// We could potentially check the display link's maximumFPS here.
+			// If it supports 120 we could force the preferredFPS to that.
+			// Currently that the's only way to drive a game at 120Hz
+		} else {
+			// Approximate frame rate: assumes device refreshes at 60 fps.
+			// Note that newer iOS devices are 120Hz screens
+			_displayLink.frameInterval = 1;
+		}
 
 		// Setup DisplayLink in main thread
-		[displayLink addToRunLoop:[NSRunLoop currentRunLoop] forMode:NSRunLoopCommonModes];
+		[_displayLink addToRunLoop:[NSRunLoop currentRunLoop]
+						   forMode:NSRunLoopCommonModes];
 	} else {
-		animationTimer = [NSTimer scheduledTimerWithTimeInterval:animationInterval target:self selector:@selector(drawView) userInfo:nil repeats:YES];
+		// This is an extremely terrible way to animate. Very low resolution with lots
+		// of drift. You would be lucky to have this called every 16ms
+		animationTimer = [NSTimer scheduledTimerWithTimeInterval:self.animationInterval
+														  target:self
+														selector:@selector(drawView)
+														userInfo:nil
+														 repeats:YES];
 	}
 
-	if (video_playing) {
-		_unpause_video();
-	}
+	// if (video_playing) {
+	// 	_unpause_video();
+	// }
 }
 
 - (void)stopAnimation {
-	if (!active)
-		return;
-	active = FALSE;
-	printf("******** stop animation!\n");
+	if (!self.isActive) return;
+
+	self.active = NO;
+	printf("stop animation!\n");
 
-	if (useCADisplayLink) {
-		[displayLink invalidate];
-		displayLink = nil;
+	if (self.useCADisplayLink) {
+		[_displayLink invalidate];
+		_displayLink = nil;
 	} else {
 		[animationTimer invalidate];
 		animationTimer = nil;
 	}
 
-	clear_touches();
-
-	if (video_playing) {
-		// save position
-	}
+	// if (video_playing) {
+	// 	_pause_video();
+	// }
 }
 
 - (void)setAnimationInterval:(NSTimeInterval)interval {
-	animationInterval = interval;
-	if ((useCADisplayLink && displayLink) || (!useCADisplayLink && animationTimer)) {
```

Comment:
This took me a few parses before I came up with the below solution. Basically in both cases we need to invalidate the current `CADisplayLink` or `NSTimer`, but only if they are actively firing. I think the code below is a bit more clear as to the intentions.